server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java

```diff
@@ -421,10 +427,9 @@ private static HitContext prepareNestedHitContext(
                     current = next;
                 }
             }
-
-            hitContext.setSourceLookup(new SourceLookup(new SourceLookup.MapSourceProvider(nestedSourceAsMap, rootSourceContentType)));
+            return new HitContext(hit, subReaderContext, nestedInfo.doc(), Source.fromMap(nestedSourceAsMap, rootSourceContentType));
         }
-        return hitContext;
+        return new HitContext(hit, subReaderContext, nestedInfo.doc(), Source.EMPTY);
```

Comment:
I think we also need lazy-loading source here? For example, we could have highlighting within `inner_hits` with `source: false` everywhere ?

I guess the test coverage is lacking here, it'd be good to add a couple cases!

Reply:
This is for a nested hit, so lazy loading wouldn't produce the correct values here anyway - we have some best effort attempts to ensure that we always load source when it's needed (we check for highlighting, for example) but I think we can improve this in a follow-up.  And more testing is always good, but this change doesn't actually change any behaviour so I don't think this shows a lack of coverage?

Reply:
Sorry for the slow response and all the back-and-forth. My concern was that before, if the source isn't eagerly loaded, then subphases could still automatically load the root document (because we pass `nestedInfo.docId()` down to `HitContext`, which gets set on its `SourceLookup`). But I couldn't find an example anywhere this is used... and as you say, it'd be weird because it wouldn't be filtered down to the relevant nested sections.

So this seems okay to me, but it could be safer to continue to fall back to lazy loading source.